### PR TITLE
Move some debug examples to the source directory

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Moves `RandomDataset`, `MNIST`, `BoringModel`, and `BoringDataModule` to `pytorch_lightning/utilities/debugging_examples.py`

Unblocks:
- https://github.com/PyTorchLightning/pytorch-lightning/pull/8715#discussion_r683478453
- https://github.com/PyTorchLightning/pytorch-lightning/pull/7614#discussion_r671183652

See the links for context and reasoning.

The use of these implementations is internal. No backwards compatibility guarantees.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/utilities/debug_examples.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+        super().__init__()
+
+        self.root = root
+        self.train = train  # training set or test set
+        self.normalize = normalize
+
+        _USE_MOCKED_MNIST = bool(os.getenv("PL_USE_MOCKED_MNIST", False))
+        if not _USE_MOCKED_MNIST:
+            # avoid users accidentally importing this MNIST implementation
+            self._torchvision_available(self.cached_folder_path)
```
Comment (carmocca) [inline pytorch_lightning/utilities/debug_examples.py:211]: All this code in this file has been just moved but this piece.

This is to error when users get this `MNIST` import accidentally with IDE autocompletion

File: tests/__init__.py [hunk 2]
```diff
@@ -34,3 +34,6 @@
     os.mkdir(_TEMP_PATH)
 
 logging.basicConfig(level=logging.ERROR)
+
+# Use our MNIST implementation on tests to avoid the torchvision dependency
+os.environ.setdefault("PL_USE_MOCKED_MNIST", "1")
```
Comment (carmocca) [inline tests/__init__.py:39]: This sets all our tests to use our MNIST implementation

File: pytorch_lightning/utilities/debug_examples.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
+# Copyright The PyTorch Lightning team.
```
Comment (carmocca) [inline pytorch_lightning/utilities/debug_examples.py:1]: Open to different filename suggestions or file locations.

Comment (Borda) [reply pytorch_lightning/utilities/debug_examples.py:1]: yep feel like the utils shall be completely inter-dependency free...
hoe about `PL/develop/samples`?

Comment (awaelchli) [reply pytorch_lightning/utilities/debug_examples.py:1]: we also have a utilities file `debugging.py` so my suggestion is to move it there or adjust naming to debugging_examples :)

Comment (carmocca) [reply pytorch_lightning/utilities/debug_examples.py:1]: > hoe about PL/develop/samples?

Are you saying `pytorch-lightning/pytorch_lightning/develop/samples` or `pytorch-lightning/develop/samples`?

The latter wouldn't work as we need it to be distributed with the source.

> debugging_examples

That's fine with me

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/debug_examples.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/utilities/debug_examples.py]: What about `pytorch_lightning/utilities/examples/...` ?

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/utilities/debug_examples.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/utilities/debug_examples.py]: IMO we want to keep this list of in-source examples as small as possible. Making it a directory might not encourage that

File: pytorch_lightning/utilities/debugging_examples.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+import random
+import time
+import urllib.request
+from typing import Optional, Tuple
+from urllib.error import HTTPError
+
+import torch
+from torch.utils.data import DataLoader, Dataset, Subset
+
+from pytorch_lightning import LightningDataModule, LightningModule
```
Comment (Borda) [inline pytorch_lightning/utilities/debugging_examples.py:30]: I am not much happy about importing these high-level structures in utils, this may lead to import cycles...

Comment (carmocca) [reply pytorch_lightning/utilities/debugging_examples.py:30]: > this may lead to import cycles...

why? our source code should **never** import this. This is just for `tests` and `pl_examples`

Comment (Borda) [reply pytorch_lightning/utilities/debugging_examples.py:30]: yes, ours not but how about users?

Comment (carmocca) [reply pytorch_lightning/utilities/debugging_examples.py:30]: I'm not sure what you mean. Users should never import these. See the top of the file

And even if they did, it wouldn't produce cyclic imports because these are only used in our tests and examples

**Review by SeanNaren — APPROVED** [review]

**Comment by Borda** [comment]
@carmocca is this still draft?

**Comment by carmocca** [comment]
> @carmocca is this still draft?

No

**Comment by carmocca** [comment]
Closing because removing the `pl_examples` directory would be a better solution. These examples currently act as an extension of tests and have very low quality as examples for new users.
